product_configurator/models/product.py

```diff
@@ -349,36 +347,57 @@ def create_variant(self, value_ids, custom_values=None):
     @api.multi
-    def value_available(self, attr_val_id, value_ids):
-        """Determines whether the attr_value from the product_template
-            is available for selection given the configuration ids and the
-            dependencies set on the product template
+    def values_available(self, attr_val_ids, sel_val_ids):
+        """Determines whether the attr_values from the product_template
+        are available for selection given the configuration ids and the
+        dependencies set on the product template
 
-            :param attr_val_id: int of product.attribute.value object
-            :param value_ids: list of attribute value ids
-
-            :returns: True or False representing availability
+        :param attr_val_ids: list of attribute values
+        :param sel_val_ids: list of attribute value ids already selected
 
+        :returns: list of available attribute values
         """
-        self.ensure_one()
-        config_lines = self.config_line_ids.filtered(
-            lambda l: attr_val_id in l.value_ids.ids
-        )
 
-        domains = config_lines.mapped('domain_id').compute_domain()
+        avail_val_ids = []
+        for attr_val_id in attr_val_ids:
 
-        for domain in domains:
-            if domain[1] == 'in':
-                if not set(domain[2]) & set(value_ids):
-                    return False
-            else:
-                if set(domain[2]) & set(value_ids):
-                    return False
-        return True
+            config_lines = self.config_line_ids.filtered(
+                lambda l: attr_val_id in l.value_ids.ids
+            )
+            domains = config_lines.mapped('domain_id').compute_domain()
+
+            # process domains as shown in this wikipedia pseudocode:
+            # https://en.wikipedia.org/wiki/Polish_notation#Order_of_operations
+            stack = []
+            for domain in reversed(domains):
+                if type(domain) == tuple:
+                    # evaluate operand and push to stack
+                    if domain[1] == 'in':
+                        if not set(domain[2]) & set(sel_val_ids):
+                            stack.append(False)
+                            continue
+                    else:
+                        if set(domain[2]) & set(sel_val_ids):
+                            stack.append(False)
+                            continue
+                    stack.append(True)
+                else:
+                    # evaluate operator and previous 2 operands
+                    # compute_domain() only inserts 'or' operators
+                    # compute_domain() enforces 2 operands per operator
+                    operand1 = stack.pop()
+                    operand2 = stack.pop()
+                    stack.append(operand1 or operand2)
+
+            # 'and' operator is implied for remaining stack elements
+            avail = True
+            while stack:
+                avail &= stack.pop()
+            if avail:
```

Comment:
@matt454357 @PCatinean 

I would love to see the block from 370-396 spun off in to its own method, `validate_domain(self, domains, sel_val_ids):` which I could then use in my defaults code which uses domains, but not from config_lines.

I think in general, this would be an ideal refactor in any case.